core/ir/aarch64/instr_create_api.h

```diff
@@ -738,6 +724,100 @@ enum {
+ * \param Rn   The input register containing the Set/Way value and cache level.
+ *             See reference manual for a detailed explanation.
+ */
+#define INSTR_CREATE_dc_cisw(dc, Rn) instr_create_0dst_1src(dc, OP_dc_cisw, Rn)
+
+/**
+ * Creates a DC CIVAC instruction to Clean and Invalidate data cache by
+ * Virtual Address to point of Coherency.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the virtual address to use.
+ *             No alignment restrictions apply to this VA.
+ */
+#define INSTR_CREATE_dc_civac(dc, Rn) instr_create_0dst_1src(dc, OP_dc_civac, Rn)
+
+/**
+ * Creates a DC CSW instruction to Clean data cache line by Set/Way.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the Set/Way value and cache level.
+ *             See reference manual for a detailed explanation.
+ */
+#define INSTR_CREATE_dc_csw(dc, Rn) instr_create_0dst_1src(dc, OP_dc_csw, Rn)
+
+/**
+ * Creates a DC CVAC instruction to Clean data cache by Virtual Address to
+ * point of Coherency.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the virtual address to use.
+ *             No alignment restrictions apply to this VA.
+ */
+#define INSTR_CREATE_dc_cvac(dc, Rn) instr_create_0dst_1src(dc, OP_dc_cvac, Rn)
+
+/**
+ * Creates a DC CVAU instruction to Clean data cache by Virtual Address to
+ * point of Unification.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the virtual address to use.
+ *             No alignment restrictions apply to this VA.
+ */
+#define INSTR_CREATE_dc_cvau(dc, Rn) instr_create_0dst_1src(dc, OP_dc_cvau, Rn)
+
+/**
+ * Creates a DC ISW instruction to Invalidate data cache line by Set/Way.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the Set/Way value and cache level.
+ *             See reference manual for a detailed explanation.
+ */
+#define INSTR_CREATE_dc_isw(dc, Rn) instr_create_0dst_1src(dc, OP_dc_isw, Rn)
+
+/**
+ * Creates a DC ICVAC instruction to Clean and Invalidate data cache by
+ * Virtual Address to point of Coherency.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the virtual address to use.
+ *             No alignment restrictions apply to this VA.
+ */
+#define INSTR_CREATE_dc_ivac(dc, Rn) instr_create_0dst_1src(dc, OP_dc_ivac, Rn)
+
+/**
+ * Creates a DC ZVA instruction to Zero data cache by Virtual Address.
+ * Zeroes a naturally aligned block of N bytes, where N is identified in
+ * DCZID_EL0 system register.
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rn   The input register containing the virtual address to use.
```

Comment:
But this is not a register: it is a memory operand, and its size is not OPSZ_sys but rather a cache line (the N above).  The docs here imply the macro will auto-convert it to a memory reference of the right size: but the test below passes in a memory reference (of the wrong size).

Similarly for all the other operations here that take in a virtual address.

Reply:
This also brings up the question of who back-aligns the address.  So this PR in its current state is mostly splitting the opcode (#4393) -- but #4400 is about making the operand reflect the actual memory range affected, which means aligning it and setting the size to the cache line size.  But xref the discussion in #4400 about how to get the cache line size for things like host != target (e.g., we run the AArch64 decoder on x86 machines).

Reply:
> and its size is not OPSZ_sys but rather a cache line

So do we add new `OPSZ_n` constants, for all known cache line size values as `n`?

> how to get the cache line size for things like host != target

From Derek's comment on #4400: "I guess we would make this a decoder setting like the ISA mode (dr_get_isa_mode()) and add routines for setting it": Can we set the cache line size for the decoder by reading it from the `TRACE_MARKER_TYPE_CACHE_LINE_SIZE` marker whenever we see it? Or is the  marker seen too late?

Reply:
For drcachesim offline use, yes we could use the marker in the trace.  But we have other uses of the decoder, and we provide drdecodelib for others to use.

From an offline discussion:
+ We could add a new OPSZ_CACHE_LINE_SIZE to avoid having to ensure we have OPSZ_ values for all possible cache line sizes.
+ opnd_size_in_bytes() would then translate OPSZ_CACHE_LINE_SIZE to a value.  For live DR it would use the already-queried line size; for drdecode we would add a new global attribute storing a line size.
+ The new global attribute would have a default of the host size and we'd add init code to get that?  Hopefully that won't be painful to share the existing code for that with drdecode.
+ Users of drdecode doing cross-arch would have to call a set routine on the global attribute to get the right size.  Maybe the default could be the most likely AArch64 value.

Reply:
> But this is not a register: it is a memory operand...

This is my understanding:
- The instruction takes a 64 bit register `<Xt>` which should contain a VA: `DC <dc_op>, <Xt>`
- The codec defines `<Xt>` as `memx0`:
`110101010000101101110100001xxxxx  n   568      dc_zva          memx0 :`
- Which creates a memory reference to the VA in `<Xt>` when decoded:
`opnd_create_base_disp(decode_reg(extract_uint(enc, 0, 5), true, false), DR_REG_NULL, 0, 0, OPSZ_sys);`

Disregarding `OPSZ_sys` for now, is that wrong?
What should it be, an absolute address using `opnd_create_abs_addr()` ?

>...and its size is not `OPSZ_sys` but rather a cache line (the N above).

This implies the DC ZVA macro should have an extra size input operand:
`#define INSTR_CREATE_dc_zva(dc, Rn, cline_sz) instr_create_1dst_1src(dc, OP_dc_zva, Rn, cline_sz)`
Which is converted to the `opnd_size_t` type.

But that introduces an input IR operand which is not part of the encoding definition. It should really be a fixed global size like the proposed `OPSZ_CACHE_LINE_SIZE`, e.g.
`opnd_create_base_disp(decode_reg(extract_uint(enc, 0, 5), true, false), DR_REG_NULL, 0, 0, opnd_size_in_bytes(OPSZ_CACHE_LINE_SIZE));`

> The docs here imply the macro will auto-convert it to a memory reference of the right size: but the test below passes in a memory reference (of the wrong size).

Just realised tests in `ir_aarch64.c` are wrong because they do an unnecessary `opnd_create_base_disp()` conversion on the register before calling `INSTR_CREATE_dc_zva()`!

Reply:
OK, it seems reasonable to keep the creation macro looking like assembly and thus taking in a bare register.  But the codec is not involved in direct synthesis of an instruction at the IR level: we need the implementation of the creation macro to turn that register into a memory operand with the cache line size (our vote is OPSZ_CACHE_LINE_SIZE).

The memory operand being a base-disp makes sense.

I think the one part we haven't thought through is the back-alignment: my comments in #4400 about back-aligning at decode time only works for an address that's a literal immediate: with a dynamic value in a register maybe the only way it could work is if it is a completely new type of operand?  Alternatively, we could leave it as a base-disp, and rely on everyone using `drutil_insert_get_mem_addr()` and having it notice the OPSZ_CACHE_LINE_SIZE and we document that that size implies back-alignment and the drutil code inserts a back-alignment (can be done w/o touching the flags on aarch64).

Reply:
We have the dr_shift_type_t and dr_extend_type_t modifiers on index operands today: we could have something like that as a modifier on the base register to back-align it, w/o requiring a whole new opnd type: just a modifier on a base-disp.

Reply:
dr_opnd_flags_t is another candidate

Reply:
Given that dr_extend_type_t is hardcoded for index registers (`opnd_get_index_extend()`) we should have something new. 

 It could be a single bit like the existing `opnd_t.value.addr.scaled` or a new `dr_base_adjust_type_t`.  There are 3 bits free for AArch64 in `opnd_t.value.addr`.  I would vote for a single bit like `opnd_t.value.addr.base_aliigned`.  It would need new setter and getter functions.

Reply:
So my vote is for:
- [ ] A new flag `opnd_t.value.addr.base_aligned` with new setter and getter functions
- [ ] A new size `OPSZ_CACHE_LINE_SIZE` with new global drdecode state for use outside of managed mode and resolution in `opnd_size_in_bytes()`
- [ ] A new codec type that sets the flag and sets `OPSZ_CACHE_LINE_SIZE`
- [ ] The creation macros turn the passed-in bare register into a base-disp that sets the flag and sets `OPSZ_CACHE_LINE_SIZE`
- [ ] A bump in the offline trace version with conditional handling of the existing workaround
- [ ] Adding support for the new flag to `drutil_insert_get_mem_addr()`

Reply:
> OK, it seems reasonable to keep the creation macro looking like assembly and thus taking in a bare register. But the codec is not involved in direct synthesis of an instruction at the IR level: we need the implementation of the creation macro to turn that register into a memory operand with the cache line size (our vote is OPSZ_CACHE_LINE_SIZE).

I've done this change but have a place holder for the memory operand size temporarily set to OPSZ_512.

There is 1 issue which needs a decision so I can implement and 1 I need to understand before implementing:

1) OPSZ_CACHE_LINE_SIZE
The Arm spec for DC ZVA says:
`Zero data cache by address. Zeroes a naturally aligned block of N bytes, where the size of N is identified in DCZID_EL0.`

`DCZID_EL0` is a system register the 4 least significant bits hold the block size:
`Log2 of the block size in words. The maximum size supported is 2KB (value == 9).`

The spec refers to block size rather than cache line size. What shall we call it?
`OPSZ_CACHE_LINE_SIZE`
`OPSZ_DC_ZVA_BLOCK_SIZE`
`Something else`

Where is the best place to initialise `OPSZ_!_SIZE`?

What should it be initialised to?
A fixed value? 2K bytes is the maximum size.
The value read from `DCZID_EL0` using `MRS <Xt>, DCZID_EL0` on startup?

2) What is the purpose of back-aligining the address? Is it a case of aligning to some previous memory boundary determined by cache line?
The Arm spec says:
`There is no alignment restriction on the address within the block of N bytes that is used.`
I guess the key word here is "_within_". We will backalign the _whole_ block to some boundary?

I noticed the `ALIGN_BACKWARD` macro called in many places. Does the proposed solution using `opnd_t.value.addr.base_aligned` flag boil down to using that macro?

Reply:
> 1. OPSZ_CACHE_LINE_SIZE

Given that other instructions like `DC GVA` and `DC GZVA` (looks like we need to add handling to treat this like `ZVA`) also use `DCZID_EL0` I would not name this with just `ZVA`.

I could see two alternatives:

A. Try to be cross-platform (maybe RISC-V has something that would need this, or some other ISA or an extension to an existing ISA we support) and use the general name `OPSZ_CACHE_LINE_SIZE`.  At init time, read `DCZID_EL0` and if it's not equal to the dcache line size then we fatally abort (would that ever happen?)

B. Have a more ISA-specific name `OPSZ_CACHE_ZERO` or sthg and set it to `DCZID_EL0`.

> 2. What is the purpose of back-aligining the address?

Tools need to know the precise memory range affected by an instruction.  `DC ZVA x0` with `x0 == my-addr` is not very useful for tools: tracers, simulator, taint tracers, etc. all want to know what actual address range is being touched.  If the IR or its provided utilities like `drutil_insert_get_mem_addr()` don't convert `my-addr` into the actual `[back-aligned-my-addr, back-aligned-my-addr + block-size)`, then every single tool has to add special code to handle this one opcode, which is not good.  From the point of view of most tools, this is just a store instruction, and we want regular things like a memory store operand to able to be handled by general code without special casing for the opcode.

> I noticed the ALIGN_BACKWARD macro called in many places. Does the proposed solution using opnd_t.value.addr.base_aligned flag boil down to using that macro?

Since the actual value of the source register is dynamic, `drutil_insert_get_mem_addr()` cannot use that macro: it will have to generate and insert raw instruction(s) to do the alignment.

Reply:
Given that `OPSZ` already implies "size" maybe `OPSZ_CACHE_LINE` is better if we go that name direction.

Re: "line" vs "block": they seem fairly interchangeable; I think the dr$sim sources also use "block" quite a bit.

Reply:
I've had a go at implementing `OPSZ_CACHE_LINE`.

The value for `OPSZ_CACHE_LINE` in `dis-a64.txt` and `ir_aarch64.expect` is currently hard-coded. I don't think any changes are required in the disassembler, but we do need to modify the harness for codec testing. Is that something that should be done for v9.0 (#5145)?

Will start on the back-alignment mods now.

Reply:
To be addressed and resolved in a post-v9.0 patch.